File: q_1175/main.cpp
#include <iostream>
#include <math.h>
using namespace std;

int primeSieve(int n)
{
    if (n < 2)
        return 0;

    bool primes[n];
    std::fill_n(primes, n, 1);
    primes[0] = 0;
    primes[1] = 0;

    for (int i = 2; i < sqrt(n); i++)
        if (primes[i] == true)
            for (int j = i * i; j < n; j += i)
                primes[j] = false;

    int count = 0;
    for (int i = 0; i < n; i++)
        count += primes[i];

    return count;
}

long factorial(int n)
{
    int mod = 1000000007;
    if (n < 2)
        return 1;

    long x = 1;
    for (int i = 1; i <= n; i++)
    {
        x *= i;
        if (x >= mod)
            x %= mod;
    }

    return x;
}

int numPrimeArrangements(int n)
{
    int mod = 1000000007;
    int s_prime, s_other;
    s_prime = primeSieve(n + 1);
    s_other = n - s_prime;

    return (factorial(s_prime) * factorial(s_other)) % mod;
}

int main()
{
    int n = 100;
    cout << numPrimeArrangements(n) << endl;
    return 0;
}

File: q_1175/main.py
import math

mod = 1000000007


def primeSieve(n: int) -> int:
    if n < 2:
        return 0

    primes = [1 for _ in range(n)]
    primes[0], primes[1] = 0, 0

    i = 2
    while i < math.sqrt(n):
        if primes[i] == 1:
            j = i * i
            while j < n:
                primes[j] = 0
                j += i
        i += 1

    return sum(primes)


def factorial(n: int) -> int:
    if n < 2:
        return 1

    x = 1
    for i in range(1, n + 1):
        x *= i
        if x >= mod:
            x %= mod
    return x


def numPrimeArrangements(n: int) -> int:
    s_prime = primeSieve(n+1)
    s_other = n - s_prime

    return (factorial(s_prime) * factorial(s_other)) % mod


if __name__ == '__main__':
    print(numPrimeArrangements(100))
